components/formats-bsd/src/loci/formats/codec/JetrawCodec.java

```diff
@@ -0,0 +1,170 @@
+import loci.formats.UnsupportedCompressionException;
+
+/**
+ */
+public class JetrawCodec extends BaseCodec {
+
+  private boolean isInitialized = false;
+
+  private byte[] convertShortArrayToByteArray(short[] shortBuffer, CodecOptions options) {
+    int pixels = options.width*options.height;
+    // convert short buffer into a byte buffer
+    int short_index = 0;
+    int byte_index = 0;
+    byte[] byteBuffer = new byte[2*pixels];
+    for(/*NOP*/; short_index != pixels; /*NOP*/) {
+      if (options.littleEndian) {
+        byteBuffer[byte_index]     = (byte) (shortBuffer[short_index] & 0x00FF);
+        byteBuffer[byte_index + 1] = (byte) ((shortBuffer[short_index] & 0xFF00) >> 8);
+      } else {
+        byteBuffer[byte_index]     = (byte) ((shortBuffer[short_index] & 0xFF00) >> 8);
+        byteBuffer[byte_index + 1] = (byte) (shortBuffer[short_index] & 0x00FF);
+      }
+      ++short_index; byte_index += 2;
+    }
+    return byteBuffer;
+  }
+
+  private String createExceptionMessage() {
+    String operatingSystem = System.getProperty("os.name").toLowerCase();
+    String message = "\n\n[JETRAW] Exception thrown"
+                   + "\n  Error loading jetraw_bioformats_plugin dynamic library."
+                   + "\n  Please verify that Jetraw is installed in default folder:";
+    if (operatingSystem.contains("win")) {
+        message = message
+                + "\n\n    C:/Program Files/Jetraw/bin64/ \n"
+                + "\n  If not, please add your Jetraw installation folder into your PATH.\n"
+                + "\n  If you need to install Jetraw, download link: "
+                + "\n    https://www.jetraw.com/download-jetraw-demo \n";
+    } else if (operatingSystem.contains("mac")) {
+        message = message
+                + "\n\n    /Applications/Jetraw UI.app/Contents/jetraw/lib/ \n"
+                + "\n  If not, please add your Jetraw installation folder"
+                + "\n  into your DYLD_FALLBACK_LIBRARY_PATH.\n"
+                + "\n  If you need to install Jetraw, download link: "
+                + "\n    https://www.jetraw.com/download-jetraw-demo \n";
+    } else if (operatingSystem.contains("nix") || operatingSystem.contains("nux") || operatingSystem.contains("aix")) {
+        message = message
+                + "\n\n    ~/jetraw/lib/ \n"
+                + "\n  If not, please add your Jetraw installation folder"
+                + "\n  into your LD_LIBRARY_PATH.\n"
+                + "\n  If you need to install Jetraw, download link: "
+                + "\n    https://www.jetraw.com/download-jetraw-demo \n";
+    }
+    return message;
+  }
+
+  private String getJetrawLibraryPath() throws MissingLibraryException {
+    String operatingSystem = System.getProperty("os.name").toLowerCase();
+    String libraryPath = "";
+    if (operatingSystem.contains("win")) {
+        libraryPath = "C:\\Program Files\\Jetraw\\bin64\\jetraw_bioformats_plugin.dll";
```

Comment:
Hard-coding paths is likely to be fairly brittle. You might consider wrapping the native libraries in a jar and then using a tool like [native-lib-loader](https://github.com/scijava/native-lib-loader) to extract the right dynamic library for the current platform. You can see an example of use in 3i's SlideBook reader at https://github.com/Intelligent-Imaging/bioformats/blob/develop/components/formats-gpl/src/loci/formats/in/SlideBook6Reader.java#L89 and info on how to then enable the related update site at https://www.intelligent-imaging.com/technical-answers

Reply:
Hi @joshmoore 

Yes, I know it is always to avoid this sort of thing. The first approach was to ship the libraries inside the .jar file so no hard-coded paths are used, but the approach was not approved. 

Our Jetraw installer by default is installed in that path in windows and the JNI library for bio-formats will come along with it. In the case the user decides to use Jetraw in another folder then the library is loaded from the PATH using loadLibrary call. 

We do not want to add more installing steps to the user and without hard-coding the default folder, I do not see how to avoid an extra step like you suggest using native-lib-loader. I could pass a json with default folders for each OS but in the end it would be the same as hard-coded paths. Not elegant at all but I do not see an alternative without adding an extra installation step, like I mentioned before. 

Please let me know your inputs @joshmoore and thanks!

Reply:
You're right that this is another installation step, but I think it's an _alternative_ one rather than an additional one. There will _always_ be 2 steps: having the right Bio-Formats version and then installing the Jetraw specific libraries. For Fiji users, having the jar I outlined above could be downloaded from an update site (as is Bio-Formats) so I think it would actually be _easier_ for a certain segment of the community.

Reply:
Thanks @joshmoore. We will discuss internally and I will come back to you. Much appreciated your contribution ;-)